src/config.f90

```diff
@@ -0,0 +1,130 @@
+module m_config
+  !! Contains all the namelist configurations used in x3d2
+  use iso_fortran_env, only: stderr => error_unit
+
+  use m_common
+
+  implicit none
+
+  type, abstract :: base_config_t
+    !! All config types have a method read to initialise their data
+  contains
+    procedure(read), deferred :: read
+  end type base_config_t
+
+  type, extends(base_config_t) :: domain_config_t
+    character(len=30) :: flow_case_name
+    real(dp) :: L_global(3)
+    integer :: dims_global(3), nproc_dir(3)
+    character(len=20) :: BC_x(2), BC_y(2), BC_z(2)
+  contains
+    procedure :: read => read_domain_nml
+  end type domain_config_t
+
+  type, extends(base_config_t) :: solver_config_t
+    real(dp) :: Re, dt
+    integer :: n_iters, n_output
+    character(3) :: poisson_solver_type, time_intg
+    character(30) :: der1st_scheme, der2nd_scheme, &
+                     interpl_scheme, stagder_scheme
+  contains
+    procedure :: read => read_solver_nml
+  end type solver_config_t
+
+  abstract interface
+    subroutine read(self, file_name, src) !&
+      import :: base_config_t
+
+      class(base_config_t) :: self
+      character(*), optional, intent(in) :: file_name
+      character(*), optional, intent(in) :: src
+    end subroutine read
+  end interface
+
+contains
+
+  subroutine read_domain_nml(self, file_name, src)
```

Comment:
just reading the definition it isn't clear what src is compared to file_name. So I'd add some comments here

Reply:
Added comments for the interface read_nml function to cover both